sdk/eventhub/event-hubs/src/receiver.ts

```diff
@@ -232,61 +264,150 @@ export class EventHubConsumer {
-        // Hence closing the link when result.length < maxMessageCount
-        await this._batchingReceiver.close();
-      }
-    } catch (err) {
-      log.error(
-        "[%s] Receiver '%s', an error occurred while receiving %d messages for %d max time:\n %O",
-        this._context.connectionId,
-        this._batchingReceiver.name,
-        maxMessageCount,
-        maxWaitTimeInSeconds,
-        err
-      );
-      throw err;
-    }
-    return result;
+    // store events across multiple retries
+    const receivedEvents: ReceivedEventData[] = [];
+
+    const retrieveEvents = (): Promise<ReceivedEventData[]> => {
+      return new Promise(async (resolve, reject) => {
+        // if this consumer was closed, _baseConsumer might be undefined.
+        // resolve the operation's promise with the events collected thus far in case
+        // the promise hasn't already been resolved.
+        if (!this._baseConsumer) {
+          return resolve(receivedEvents);
+        }
+
+        let timer: any;
+        const logOnAbort = (): void => {
+          const baseConsumer = this._baseConsumer;
+          const name = baseConsumer && baseConsumer.name;
+          const address = baseConsumer && baseConsumer.address;
+          const desc: string =
+            `[${this._context.connectionId}] The request operation on the Receiver "${name}" with ` +
+            `address "${address}" has been cancelled by the user.`;
+          log.error(desc);
+        };
+
+        const rejectOnAbort = async (): Promise<void> => {
+          logOnAbort();
+          try {
+            await this.close();
+          } finally {
+            return reject(new AbortError("The receive operation has been cancelled by the user."));
+          }
+        };
+
+        // operation has been cancelled, so exit immediately
+        if (abortSignal && abortSignal.aborted) {
+          return await rejectOnAbort();
+        }
+
+        const onAbort = (): void => {
+          clearTimeout(timer);
+          rejectOnAbort();
+        };
+
+        // updates the prefetch count so that the baseConsumer adds
+        // the correct number of credits to receive the same number of events.
+        const prefetchCount = Math.max(maxMessageCount - receivedEvents.length, 0);
+        if (prefetchCount === 0) {
+          return resolve(receivedEvents);
+        }
+
+        log.batching(
+          "[%s] Receiver '%s', setting the prefetch count to %d.",
+          this._context.connectionId,
+          this._baseConsumer && this._baseConsumer.name,
+          prefetchCount
+        );
+
+        const cleanUpBeforeReturn = (): void => {
+          if (this._baseConsumer) {
+            this._baseConsumer.clearHandlers();
+          }
+          if (abortSignal) {
+            abortSignal.removeEventListener("abort", onAbort);
```

Comment:
Why shouldnt this be inside `clearHandlers`?

Reply:
There isn't a way to remove all event listeners from an `abortSignal`, you need to have a reference to the function passed in when adding an event listener to remove it. We could pass the `onAbort` function to the underlying receiver and store it like we do `_onMessage` and `_onError`.

Originally when I was handling the link creation here as well, I needed to be able to add and remove the listener in this function. Now that link creation happens in `registerHandlers` it should work if the EventHubReceiver handles adding and removing the listener.